src/adaptors/amqp/qd_connector.c

```diff
@@ -439,27 +500,90 @@ void qd_connector_remove_connection(qd_connector_t *connector, bool final, const
-        // TBD
-        qd_log(LOG_SERVER, QD_LOG_DEBUG,
-               "Connector %s new ordinal: %"PRIu64", previous: %"PRIu64,
-               ctor_config->config.name, new_ordinal, ctor_config->tls_ordinal);
-        ctor_config->tls_ordinal = new_ordinal;
-    }
+    // Expect: this callback is only registered for inter-router connectors (for now)
+    assert(strcmp(ctor_config->config.role, "inter-router") == 0);
+
+    // destroy all connections whose connectors use an expired TLS ordinal:
 
     if (new_oldest_ordinal > ctor_config->tls_oldest_valid_ordinal) {
-        // TBD
+
         qd_log(LOG_SERVER, QD_LOG_DEBUG,
-               "Connector %s new oldest valid ordinal: %"PRIu64", previous: %"PRIu64,
+               "Connector %s new oldest valid TLS ordinal: %"PRIu64", previous: %"PRIu64,
                ctor_config->config.name, new_oldest_ordinal, ctor_config->tls_oldest_valid_ordinal);
+
+        // Spin through all connectors and delete those with an expired ordinal
+
+        qd_connector_t *ctor = DEQ_HEAD(ctor_config->connectors);
+        while (ctor) {
+            if (ctor->tls_ordinal < new_oldest_ordinal) {
+                qd_connector_t *delete_me = ctor;
+                ctor = DEQ_NEXT(ctor);
+                DEQ_REMOVE(ctor_config->connectors, delete_me);
+                qd_connector_close(delete_me);
+                qd_connector_decref(delete_me);
+            } else {
+                ctor = DEQ_NEXT(ctor);
+            }
+        }
+
         ctor_config->tls_oldest_valid_ordinal = new_oldest_ordinal;
     }
+
+    if (new_ordinal > ctor_config->tls_ordinal) {
+
+        qd_log(LOG_SERVER, QD_LOG_DEBUG,
+               "Connector %s new TLS ordinal: %"PRIu64", previous: %"PRIu64,
+               ctor_config->config.name, new_ordinal, ctor_config->tls_ordinal);
+
+        // Prevent existing connections that use older TLS ordinals from reconnecting. The goal is to allow older (but
+        // still valid) TLS sessions to pass existing flows but avoid sending new flows over them.
+
+        qd_connector_t *ctor = DEQ_HEAD(ctor_config->connectors);
+        while (ctor) {
+            bool active;
+            qd_connector_disable_reconnect(ctor, &active);
+            if (!active) {
+                // If the connector does not have an active connection it can be deleted
+                qd_connector_t *delete_me = ctor;
+                ctor = DEQ_NEXT(ctor);
+                DEQ_REMOVE(ctor_config->connectors, delete_me);
+                qd_connector_close(delete_me);
+                qd_connector_decref(delete_me);
+            } else {
+                ctor = DEQ_NEXT(ctor);
+            }
+        }
+
+        // Create a new set of connectors that will use the updated TLS credentials
+
+        ctor_config->tls_ordinal = new_ordinal;
+
+        for (int i = 0; i < ctor_config->data_connection_count; i++) {
+            ctor = qd_connector_create(ctor_config, true);
```

Comment:
Can we please create two new functions `qd_connector_create_data_conn()` and `qd_connector_create_control_conn()` and remove the boolean parameter ?

Reply:
I like this idea but when I was looking into making that change it didn't make sense because qd_connector_create() can create other types of connectors as well (edge, inter-edge, route-container, normal).  All this is hidden in the qd_connector_create() constructor, which is good.

The flag is a hack that basically tells the constructor "ignore the configured role and use inter-router-data instead".

Given all this I'd prefer not to have a dedicated qd_connector_create_data_conn() because that would require duplicating all of qd_connector_create() and just leave out a few lines of data connection specific logic.

Or maybe instead of a true/false parameter we define an enum "ROLE_OVERRIDE_DATA_CONN" and pass that?

Reply:
Gah, I'm stumped.

I ended up moving all the connector creation logic to a local helper function and commented that. Puts it all in one place for now.  Let me know what you think.